src/main/java/org/zalando/nakadi/throttling/SampledTotal.java

```diff
@@ -0,0 +1,41 @@
+package org.zalando.nakadi.throttling;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class SampledTotal {
```

Comment:
I would implement it a little bit differently, though your implementation is cool. I would store the tuple of three counts (to have more consistent sums).

The basic idea, is that your SampleCollector can look something like:
```java
SampleCollector {
    ReadWriteLock rwl;

    CounterBin currentBin;
    long currentBinSwitchTime;

    HashMap<Long, CounterBin> previousBins;
    long previousBinsSwitchTime;
    ...

}
```

And `currentBin` contains your three [`LongAdder`](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/LongAdder.html) values for collecting the values.

You 1) ReadLock, 2) If current time is higher, then `currentBinSwitchTime`, you 3) WriteLock, 4) create a new `currentBin` 4) add old bin to the map (actually it can be a simple `HashMap` in this case), 5) proclaim new `currentBin` and update the `currentBinSwitchTime` 5) release the locks.

See http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html for examples in case have needed before.

This operation will be done only once per `windowDuration` time and lock for the time of the swap and map update.

Clean up of the previous bins can be done using the same approach, by swapping it away during the `currentBin` swap operation or from a separate thread.

Just my 2 cents :D